### PR TITLE
WPS303: Allow underscore with 3 digits on the right

### LINKED ISSUE #3120: Rework `WPS303`: it should only be possible to write `_` with 3 digits on the right
Right: `1_000` and `10_000` and `100_000` and `1_000_000`
Wrong: `1_00` and `10_00`

### PR DESCRIPTION
**Pull request by mmmylnikov**
## Checklist

- [+] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [+] I have created at least one test case for the changes I have made
- [+] I have updated the documentation for the changes I have made
- [+] I have added my changes to the `CHANGELOG.md`

## Related issues

- Closes #3120

This is my first pull request to this project, so please bear with me if I made any mistakes. Any comments or suggestions are welcome.

### CODE REVIEW

File: tests/test_visitors/test_tokenize/test_primitives/test_numbers/test_underscored_numbers.py [hunk 1]
```diff
@@ -39,7 +42,19 @@ def test_underscored_number(
...
+    '+1_000',
+    '+1_000.0',
+    '-1_000',
+    '-1_000.0',
+    '10_000',
+    '100_000',
+    '100_000_000',
+    '"1_00"',
     '"10_00"',
+    '"100_00"',
```
Comment (sobolevn) [inline tests/test_visitors/test_tokenize/test_primitives/test_numbers/test_underscored_numbers.py:56]: This is not correct. It should be `10_000`

File: wemake_python_styleguide/violations/consistency.py [hunk 2]
```diff
@@ -266,21 +266,25 @@ class UnderscoredNumberViolation(TokenizeViolation):
...
         million = 1000000
+        formatted_million = 1_000_000
+        big_number = 10_000_000_000_000
 
         # Wrong:
         phone = 8_83_134_43
         million = 100_00_00
+        big_number = 100000_0000_000
 
     .. versionadded:: 0.1.0
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/consistency.py:289]: Please, add `.. versionchanged:: 1.0.0` with an explanation: what has changed.

File: wemake_python_styleguide/visitors/tokenize/primitives.py [hunk 3]
```diff
@@ -23,6 +23,15 @@ def _replace_braces(string: str) -> str:
     return string.lstrip("'").rstrip("'")
 
 
+def _is_wrongly_underscored_number(string: str) -> bool:
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/tokenize/primitives.py:26]: Please, move this to `logic/`

File: tests/test_visitors/test_tokenize/test_primitives/test_numbers/test_underscored_numbers.py [hunk 4]
```diff
@@ -39,7 +42,19 @@ def test_underscored_number(
     '1000.0',
     '-333555',
     '-333555.5',
+    '1_000',
+    '1_000.0',
+    '+1_000',
+    '+1_000.0',
+    '-1_000',
+    '-1_000.0',
```
Comment (sobolevn) [inline tests/test_visitors/test_tokenize/test_primitives/test_numbers/test_underscored_numbers.py:50]: There are also cases like `1.0_0`, please test them.

**Comment by mmmylnikov** [comment]
@sobolevn, thank you for your feedback. In return, I am pleased to inform you that I have made the changes you requested.

However, I have a few questions or comments:

1) Did I correctly understand that the case of numbers indicating a version:

https://github.com/mmmylnikov/wemake-python-styleguide/blob/d054c3dcfd4b501b0ae557611c8874fbd8de0afb/tests/test_visitors/test_tokenize/test_primitives/test_numbers/test_underscored_numbers.py#L22

- `1.0_0` should be used in the "test_underscored_number" test (to identify as incorrect underscoring)
- `"1.0_0"` should be used in the "test_correct_number" test (to identify as acceptable if specified as a string)

    
2) I am not sure if I correctly moved the `is_wrongly_underscored_number` function to 

https://github.com/mmmylnikov/wemake-python-styleguide/blob/d054c3dcfd4b501b0ae557611c8874fbd8de0afb/wemake_python_styleguide/logic/tokens/strings.py#L35

File: tests/test_visitors/test_tokenize/test_primitives/test_numbers/test_underscored_numbers.py [hunk 5]
```diff
@@ -44,7 +52,24 @@ def test_underscored_number(
...
+        '100_000',
+        '100_000_000',
+        '"1_00"',
+        '"10_000"',
+        '"100_000"',
+        '"-100_000"',
+        '"1.0_0"',
+        '"3_3.3"',
+        '"1_0_0"',
+        '"1_0_0_1"',
```
Comment (sobolevn) [inline tests/test_visitors/test_tokenize/test_primitives/test_numbers/test_underscored_numbers.py:71]: These numbers are not correct. Number can only considered correct if after split on `_`:
- First group has length of `1`, `2` or `3`
- All other groups have length of strictly `3` 

The same with `float` numbers: we split on `.` and apply this rule to the `float` part as well:
- `0.0_005` is correct
- `100_000.000_001` is correct
- `0.00_05` is not correct

File: tests/test_visitors/test_tokenize/test_primitives/test_numbers/test_underscored_numbers.py [hunk 6]
```diff
@@ -11,11 +11,19 @@
...
-        '333_555',
+        '1_00',
+        '10_00',
+        '100_00',
+        '-1_0',
+        '-10_0',
+        '-100_0',
+        '1000_000',
+        '-1000_000',
+        '1.0_0',
```
Comment (sobolevn) [inline tests/test_visitors/test_tokenize/test_primitives/test_numbers/test_underscored_numbers.py:22]: Please increase coverage of invalid `float` numbers. Add: `0_0.00_1` and `0.000_1`

File: tests/test_visitors/test_tokenize/test_primitives/test_numbers/test_underscored_numbers.py (hunk 5 above)
Comment (Yurgers) [reply tests/test_visitors/test_tokenize/test_primitives/test_numbers/test_underscored_numbers.py:71]: I will take part in your discussion.

is the number "1.0_0" considered a number?
or should it be interpreted as a string value, due to the fact used that there are quotes in quotes.


I was earlier to see a used shortened SemVer, an example: `version = "2.4_153"`.
and it was just a string field.

Comment (sobolevn) [reply tests/test_visitors/test_tokenize/test_primitives/test_numbers/test_underscored_numbers.py:71]: Oh, sorry. I missed these quotes. Please, add a comment that you also test strings.

**Comment by mmmylnikov** [comment]
@sobolevn , sorry for the mistake. I misunderstood the problem with floating point numbers. I've made edits and added more test cases.

**Review by sobolevn — APPROVED** [review]
Thank you! Congrats on your first PR :)

**Pull request merged**